### PR TITLE
🎅🎄🎁

### LINKED ISSUE #146: (Community Discussion) Run multiply PHP versions at the same time
# Community Discussion: multiple PHP versions

Hi everybody,

hope you enjoy the devilbox so far. I was thinking about a new feature, which might be a bit difficult to get it done, could however be something very useful for some of you:

## Feature

**Be able to run multiple PHP-FPM container at the same time**

## Benefits

- test your site simultaneously with different PHP versions
- useful for migrating from one version to another
- useful for continuous integration tests against multiple versions

## Feedback

Before I am going to wrap my head around this, I wanted to check back with you guys, If it is actually any useful for you. So if this is of any use for you, let me know. Also if you have something similar in mind, just give me a comment here.

### PR DESCRIPTION
**Pull request by cytopia**
## Release v3.0.0-beta-0.1 (2022-12-24) 🎅🎄🎁

* ### Run different PHP versions per project
* ### Reverse Proxy automation

Please report issues in **this PR** or reach out in **[Discord](https://discord.gg/2wP3V6kBj4)**.

Screenshot 2022-12-25 18-05-25  selection

---

This is a beta release, using a completely rewritten set of HTTPD server, based on these changes:

* https://github.com/devilbox/docker-nginx-stable/pull/55
* https://github.com/devilbox/docker-nginx-mainline/pull/57
* https://github.com/devilbox/docker-apache-2.2/pull/53
* https://github.com/devilbox/docker-apache-2.4/pull/54

Once it has been tested by the community, and potential errors have been addressed, a new major version will be released.

**IMPORTANT:** This release required you to copy `env-example` over onto `.env` due to some changes in variables.

### TL;DR

1. **Multiple PHP Versions**<br/>
    Here is an example to run one project with a specific PHP version<br/>
    ```bash
    # Enable all PHP versions
    cp compose/docker-compose.override.yml-php-multi.yml docker-compose.override.yml
    # Start default set and php80
    docker-compose up php httpd bind php80
    ```
    file: `/shared/httpd/<project>/.devilbox/backend.cfg`
    ```
    conf:phpfpm:tcp:php80:9000
    ```
2. **Automated Reverse Proxy setup**<br/>
    Here is an example to proxy one project to a backend service (e.g. NodeJS or Python application, which runs in the PHP container on port 3000)<br/>
    file: `/shared/httpd/<project>/.devilbox/backend.cfg`
    ```
    conf:rproxy:http:127.0.0.1:3000
    ```
#### PHP hostnames and IP addresses

**Note:** Use IP addresses instead of hostnames for Nginx

| PHP Version | Hostname | IP address     |
|-------------|----------|----------------|
| 5.4         | php54    | 172.16.238.201 |
| 5.5         | php55    | 172.16.238.202 |
| 5.6         | php56    | 172.16.238.203 |
| 7.0         | php70    | 172.16.238.204 |
| 7.1         | php71    | 172.16.238.205 |
| 7.2         | php72    | 172.16.238.206 |
| 7.3         | php73    | 172.16.238.207 |
| 7.4         | php74    | 172.16.238.208 |
| 8.0         | php80    | 172.16.238.209 |
| 8.1         | php81    | 172.16.238.210 |
| 8.2         | php82    | 172.16.238.211 |

~~### Important~~

~~Currently the Intranet Command and Control Center is not yet self sufficient as it requires a supervisord configuration file.~~

~~In the root of the Devilbox repository, add `supervisorctl.conf` into `supervisor/` directory. **Important:** It must have this name.~~
~~```ini~~
~~[supervisorctl]~~
~~serverurl=http://httpd:9001~~
~~username=supervisord~~
~~password=mypassword~~
~~```~~

### Fixed
- Fixed Protocol substitution bug in Reverse Proxy generation for Apache 2.2 and Apache 2.4 [vhost-gen #49](https://github.com/devilbox/vhost-gen/pull/49) [vhost-gen #50](https://github.com/devilbox/vhost-gen/pull/50)
- Fixed missing module `mod_proxy_html` in Apache 2.4 as per requirement from `vhost-gen` for Reverse Proxy setup
- Fixed encoding issue with Apache 2.4 Reverse Proxy by enabling `mod_xml2enc` module (Required by `mod_proxy_html`)
- Allow to run different PHP versions per project. fixes [#146](https://github.com/cytopia/devilbox/issues/146)

### Added
- New HTTPD server capable of auto reverse proxy creation (and different PHP versions per project)
- Intranet: Added Command & Control center to view watcherd logs and retrigger config in case of vhost changes
- Environment variable `DEVILBOX_HTTPD_MGMT_PASS`
- Environment variable `DEVILBOX_HTTPD_MGMT_USER`
- New Docker Compose Override file `docker-compose.override.yml-php-multi.yml` (allows to run multiple PHP versions).
- Update Bind to latest version

### Changed
- Disabled `psr` extension by default [php-psr #78](https://github.com/jbboehr/php-psr/issues/78#issuecomment-722290110)
- Disabled `phalcon` extension by default
- Environment variable `DEBUG_COMPOSE_ENTRYPOINT` renamed to `DEBUG_ENTRYPOINT`
- Environment variable `HTTPD_TIMEOUT_TO_PHP_FPM` renamed to `HTTPD_BACKEND_TIMEOUT`

### CODE REVIEW

**Comment by llaville** [comment]
Merry Christmas @cytopia, and thanks for this gift !

One little question, about hostnames : is it a typo when you write the same hostname for PHP versions, or expected results ?

<html><body>


5.4 | php54 | 172.16.238.201
-- | -- | --
5.5 | php54 | 172.16.238.202
5.6 | php54 | 172.16.238.203


</body>
</html>

**Comment by cytopia** [comment]
> Merry Christmas @cytopia, and thanks for this gift !
> 
> One little question, about hostnames : is it a typo when you write the same hostname for PHP versions, or expected results ?
> 
> 5.4	php54	172.16.238.201
> 5.5	php54	172.16.238.202
> 5.6	php54	172.16.238.203

Thanks and fixed :-)

**Comment by cytopia** [comment]
Now also showing the configured Backend on the vhosts.php page in the intranet.

**Comment by minhchu** [comment]
So much awesome, thank you

**Comment by razorcell** [comment]
The box is cool Thanks for sharing this fantastic work !

**Comment by cw95403** [comment]
I am working on getting different versions of PHP running with different projects.

The error I am getting is the following when attempting to access a project that uses other than the default version of PHP: 
Service Unavailable
The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.

I am on an M1 Mac with:
Devilbox v3.0.0-beta-0.3 (2023-01-02)

I have run through the configuration instructions on https://github.com/cytopia/devilbox/pull/942

docker-compose.override.yml is active
PHP56 and PHP81 are loaded.  PHP81 is the system default
backend.cfg consists of: conf:phpfpm:tcp:php56.local:9000

The Devilbox virtual hosts page displays the following in the backend column: tcp://php56.local:9000

Any suggestions on how to get it working would be much appreciated.

backend.cfg
`conf:phpfpm:tcp:php56.local:9000`

.env
DEBUG_ENTRYPOINT=2
DOCKER_LOGS=0
DEVILBOX_PATH=.
LOCAL_LISTEN_ADDR=
TLD_SUFFIX=local
EXTRA_HOSTS=devilbox.loc=
NEW_UID=502
NEW_GID=20
TIMEZONE=America/Los_Angeles
DNS_CHECK_TIMEOUT=1
DEVILBOX_UI_SSL_CN=localhost,*.localhost,devilbox,*.devilbox,httpd
DEVILBOX_UI_PROTECT=0

**Comment by jhaveDK** [comment]
@cytopia Nice work, but its possible to add a function so its only starts the PHP versions there are in use with "docker-compose up -d" ?

**Comment by zulfiqarpk92** [comment]
I am trying to setup multiple php versions. But it is giving "File not found" error in browser when I setup a project with specific php version e.g php81. I have added backend.cfg for the project like below.

`conf:phpfpm:tcp:172.16.238.210:9000`

Any idea what I am doing wrong?

**Comment by vikas5914** [comment]
@zulfiqarpk92 Can you please provide where you put the file as well the file name? 

Also, provide a screenshot of Vhost in Devilbox localhost.

Make sure you restart the Devilbox again. sometimes it does not take the first restart.

**Comment by zulfiqarpk92** [comment]
@vikas5914 I have placed the file in   `<project>/.devilbox/backend.cfg`. Check below screenshot. It is from the main php container.

image

Screenshot for Vhosts:

image

I have restarted several times but no luck.

Thanks

**Comment by zulfiqarpk92** [comment]
@vikas5914 I have placed the file in   `<project>/.devilbox/backend.cfg`. Check below screenshot. It is from the main php container.

image

Screenshot for Vhosts:

image

I have restarted several times but no luck.

Thanks

**Comment by vikas5914** [comment]
@zulfiqarpk92 Can you try `conf:phpfpm:tcp:php80:9000`

**Comment by zulfiqarpk92** [comment]
@vikas5914 I found the issue. It was not the IP. Actually, I have my projects placed in a separate directory and created vhosts using soft links in the dvlbox data/www directory. In main compose file I had mounted the projects directory but not in the other php containers that's why it was giving file not found error. It started working after adding volumes to the php containers.

Thanks

**Comment by frankbret** [comment]
Be really careful with the "/shared/httpd/<project>/.devilbox/backend.cfg" file to NOT LEAVE BLANK LINE at the end of the file. Else default php run and Devilbox Comand & Control "watcherd stderr" said :
[WARN] [abc.proweb.local] Backend rewrite config is invalid: Invalid backend conf: in: 'conf:phpfpm:tcp:php74:9000
'. It must be a valid port
[WARN] [abc.proweb.local] Backend rewrite: skipping

**Comment by wnunezc** [comment]
Hello, I am interested in something in particular.

If I have two projects in their respective directories of my windows (Host OS)...
I want them both to use php 7.4.

Is it possible to mount these projects as if they were volumes and that both use the same php7.4 to save resources?
